stuffer/s2n_stuffer_network_order.c

```diff
@@ -0,0 +1,185 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+#include "error/s2n_errno.h"
+
+#include "stuffer/s2n_stuffer.h"
+
+#include "utils/s2n_safety.h"
+
+#define SIZEOF_UINT24 3
+
+static int s2n_stuffer_write_network_order(struct s2n_stuffer *stuffer, uint32_t input, uint8_t length)
+{
+    notnull_check(stuffer);
+    GUARD(s2n_stuffer_skip_write(stuffer, length));
+    uint8_t *data = stuffer->blob.data + stuffer->write_cursor - length;
+
+    for (int i = 0; i < length; i++) {
+        uint8_t shift = (length - i - 1) * 8;
+        data[i] = (input >> (shift)) & 0xFF;
+    }
+
+    return S2N_SUCCESS;
+}
+
+int s2n_stuffer_read_uint8(struct s2n_stuffer *stuffer, uint8_t * u)
+{
+    GUARD(s2n_stuffer_read_bytes(stuffer, u, 1));
+
+    return 0;
+}
+
+int s2n_stuffer_write_uint8(struct s2n_stuffer *stuffer, const uint8_t u)
+{
+    GUARD(s2n_stuffer_write_bytes(stuffer, &u, 1));
+
+    return 0;
+}
+
+int s2n_stuffer_read_uint16(struct s2n_stuffer *stuffer, uint16_t * u)
+{
+    uint8_t data[2];
+
+    GUARD(s2n_stuffer_read_bytes(stuffer, data, sizeof(data)));
+
+    *u = data[0] << 8;
+    *u |= data[1];
+
+    return 0;
+}
+
+int s2n_stuffer_write_uint16(struct s2n_stuffer *stuffer, const uint16_t u)
+{
+    return s2n_stuffer_write_network_order(stuffer, u, sizeof(u));
+}
+
+int s2n_stuffer_reserve_uint16(struct s2n_stuffer *stuffer, struct s2n_stuffer_reservation *reservation)
+{
+    notnull_check(stuffer);
+    notnull_check(reservation);
+
+    reservation->stuffer = stuffer;
+    reservation->write_cursor = stuffer->write_cursor;
+    reservation->length = sizeof(uint16_t);
+
+    GUARD(s2n_stuffer_skip_write(stuffer, reservation->length));
+    memset_check(stuffer->blob.data + reservation->write_cursor, S2N_WIPE_PATTERN, reservation->length);
+
+    return S2N_SUCCESS;
+}
+
+int s2n_stuffer_read_uint24(struct s2n_stuffer *stuffer, uint32_t * u)
+{
+    uint8_t data[3];
+
+    GUARD(s2n_stuffer_read_bytes(stuffer, data, sizeof(data)));
+
+    *u = data[0] << 16;
+    *u |= data[1] << 8;
+    *u |= data[2];
+
+    return 0;
+}
+
+int s2n_stuffer_write_uint24(struct s2n_stuffer *stuffer, const uint32_t u)
+{
+    return s2n_stuffer_write_network_order(stuffer, u, SIZEOF_UINT24);
+}
+
+int s2n_stuffer_read_uint32(struct s2n_stuffer *stuffer, uint32_t * u)
+{
+    uint8_t data[4];
+
+    GUARD(s2n_stuffer_read_bytes(stuffer, data, sizeof(data)));
+
+    *u = ((uint32_t) data[0]) << 24;
```

Comment:
Is there a reason there is a utility function for write, but not read?

Reply:
I'm intentionally not modifying any of the read methods. I'm just copying them over as-is. We can factor out their common logic in another PR.

Reply:
Reasonable.